lib/ansible/module_utils/basic.py

```diff
@@ -945,16 +946,19 @@ def set_default_selinux_context(self, path, changed):
         context = self.selinux_default_context(path)
         return self.set_context_if_different(path, context, False)
 
-    def set_context_if_different(self, path, context, changed, diff=None):
+    def set_context_if_different(self, path, context, changed, diff=None, expand=True):
 
         if not HAVE_SELINUX or not self.selinux_enabled():
             return changed
-        cur_context = self.selinux_context(path)
+        if expand:
+            path = os.path.expanduser(os.path.expandvars(path))
+        b_path = to_bytes(path, errors='surrogate_or_strict')
+        cur_context = self.selinux_context(b_path)
```

Comment:
I believe selinux uses native strings (byte strings in python2 and text strings in python3) rather than always using byte strings.  So that's why we weren't using to_bytes here earlier.  We may need to move the to_native call earlier, though.  I'm not sure if it was all selinux functions or only some of them which had bugs if the wrong type of string was passed to them.